### ISSUE TITLE
Resolve start_url against manifest URL

### ISSUE THREAD

**Issue by benfrancis** [w3c/manifest#330]
Proposal: If manifest URL must be same-origin with document URL and document URL must be same-origin with start_url, then can we simplify the steps for processing a start_url by simply resolving it against manifest URL, without needing to make reference to document URL?

**Comment by marcoscaceres** [comment]
On March 13, 2015 at 9:42:45 AM, Ben Francis (notifications@github.com) wrote:
> Proposal: If manifest URL must be same-origin with document URL and document URL must  
> be same-origin with start_url, then can we simplify the steps for processing a start_url  
> by simply resolving it against manifest URL, without needing to make reference to document  
> URL?

Yep, I think that would be fine.

**Comment by kenchris** [comment]
It seems that MS wants to use the manifest to add apps to their store. I don't know how that will fit in this scenario and how to fix it. @anssiko @alxlu

We are getting similar request for Crosswalk

**Comment by anssiko** [comment]
AFAIK, in the store scenario, the `start_url` URL would not be same-origin with the manifest URL if the manifest is hosted on `yet-another-app-store.com` while the `start_url` points to `awesome-app.com`.

Similarly, in the Crosswalk use case, the manifest URL could be `app://uuid`. Cf. a glorified bookmark that sits on the local filesystem and when navigated to initiates the installation process.

**Comment by kenchris** [comment]
that works for packaged apps, but not hosted ones

**Comment by jmajnert** [comment]
I guess app stores will need to digest manifests hosted within application's origin. The problem here is that there's no mechanism for telling the UA "hey, now install the app whose manifest is here: https://myapp.com/manifest.json"
So the app store would need to redirect the UA to the apps start page (possibly start_url) so that the manifest is ingested from the <link> element.

**Comment by benfrancis** [comment]
The current manifest specification does not lend itself well to use by traditional app stores, but @jmajnert is right... the problem isn't that the manifest can't be hosted by the store (the store can can just index the manifest hosted by the app as the Firefox Marketplace does), it's that there's no installation API which an app store can use to trigger the installation of an app.

The current specification assumes that the user navigates to the web app and uses it first, then optionally installs it later from browser UI. Mozilla has a proprietary installation API [1] which is currently used by the Firefox Marketplace to install Firefox Apps, but for W3C web apps it will need to replace the "install" button with an "open" button. The user can open the app in the browser to use it instantly, then optionally install it to break it out of the browser and use it as a standalone app. As I understand it the Chrome implementation will use heuristics to determine when to prompt the user to install the app while they are using it in the browser.

So far it has been an intentional decision to not have an installation API in order to prevent install buttons popping up all over the web, but if there's an unfulfilled need that some implementers have then it would be great to hear about those use cases in more detail. If there's interest in standardising an installation API for example then I would suggest opening a new issue to discuss that so we can gather use cases.

1. https://developer.mozilla.org/en-US/docs/Web/API/DOMApplicationsRegistry/install

**Comment by jmajnert** [comment]
> So far it has been an intentional decision to not have an installation API in order to prevent install buttons popping up all over the web, but if there's an unfulfilled need that some implementers have then it would be great to hear about those use cases in more detail. If there's interest in standardising an installation API for example then I would suggest opening a new issue to discuss that so we can gather use cases.

Right! As said before - this spec already allows for creating app stores based on the manifest. The installation API is IMHO not indispensable.

**Comment by alxlu** [comment]
> It seems that MS wants to use the manifest to add apps to their store. I don't know how that will fit in this scenario and how to fix it. @anssiko @alxlu

I'm not sure how well it fits into the scenario, but it looks like something a Store would be able to support. @kenchris, how is Crosswalk planning on handling the submission process for hosted apps?

**Comment by kenchris** [comment]
Crosswalk packaged the apps on some platforms like Android. As we don't control the store, there are a few other considerations.

**Comment by kenchris** [comment]
What I think that we will do:

1. Allow packaging of local apps (from local source repo) which fulfills the requirement of start_url being same origin
2. Allow packaging "hosted apps" by pointing our packaging tool to an url containing a manifest.json or referring to one.
3. Allow packaging "hosted apps" by supplying a separate manifest.json (thus, not fulfilling the start_url requirement). In this case we will only consider the start_url, title and icon fields, and apps will basically launch in an existing browser (ie display-mode: browser). Basically a regular link to a web site.

**Comment by anssiko** [comment]
I think we haven't mentioned anything about "packaging" in the spec to date, since the feature is out of scope. We've been careful in targeting the Web as defined by the HTML spec and friends. http://w3ctag.github.io/packaging-on-the-web/ is an attempt to make "packaging" part of the Web, but it is early.

One way forward would be to define the "packaging extensions" separately, and in the extension spec patch the core spec where needed. We could add new extension points to the core spec if needed, to make patching less painful.

**Comment by kenchris** [comment]
For hosted apps, that it would actually be packaged or not, is just an implementation detail.

I mean "Allow packaging 'hosted apps' by pointing our packaging tool to an url containing a manifest.json or referring to one." could be

"Allow adding 'hosted apps' to store by pointing the store "add an app" tool to an url containing a manifest.json or referring to one."

**Comment by alxlu** [comment]
>For hosted apps, that it would actually be packaged or not, is just an implementation detail.

This was our understanding as well.

**Comment by anssiko** [comment]
I suggest we collect the requirements to "allow packaging 'hosted apps'" to this issue. With the requirements at hand, we can craft an extensions spec, and see whether some of it should be upstreamed to the core without breaking the Web's security model.

@alxlu @kenchris Do you have requirements in mind you'd like to share? Is this a valid requirement: https://github.com/w3c/manifest/issues/330#issuecomment-82251195

**Comment by kenchris** [comment]
I am not sure it is valid, such a store could very well link directly to the manifest.json hosted on the actual site, and could still get information shown on the store from that same file.

Allowing third-parties to change icon, title, orientation, launch screen etc, seems wrong. Like the store could show ads for other services and it would seem like that is associated with the app, which it might not be.

**Comment by PaulKinlan** [comment]
Just to clarify what is happening.  Are we saying the manifest must be on the same domain as page it is linked from?  If so that is directly the opposite of what we agreed a while back (can't recall the issue number).

**Comment by jmajnert** [comment]
>Just to clarify what is happening. Are we saying the manifest must be on the same domain as page it is linked from? If so that is directly the opposite of what we agreed a while back (can't recall the issue number).

Yup: https://github.com/w3c/manifest/issues/272#issuecomment-78308146

**Comment by marcoscaceres** [comment]
Depends on #272. Hopefully resolve all this on Wednesday once and for all.
